crates/polars-stream/src/nodes/io_sources/parquet/init.rs

```diff
@@ -120,122 +114,78 @@ impl ParquetSourceNode {
+            }
+            PolarsResult::Ok(())
+        });
 
-                    dfs.reserve(n);
-                    dfs.extend(iter);
+        // Distributes morsels across pipelines. This does not perform any CPU or I/O bound work -
+        // it is purely a dispatch loop.
+        let distribute_task = io_runtime.spawn(async move {
+            let mut morsel_seq = MorselSeq::default();
+            while let Some(decode_fut) = decode_recv.recv().await {
+                let df = decode_fut.await?;
+                if df.is_empty() {
+                    continue;
                 }
 
-                let mut df = dfs.pop_front().unwrap();
-                let morsel_seq = *morsel_seq_ref;
-                *morsel_seq_ref = morsel_seq.successor();
-
-                loop {
-                    use crate::async_primitives::connector::SendError;
-
-                    let channel_index = indexed_wait_group.index();
-                    let wait_token = indexed_wait_group.token();
-
-                    match raw_morsel_senders[channel_index].try_send((df, morsel_seq, wait_token)) {
-                        Ok(_) => {
-                            wait_groups.push(indexed_wait_group.wait());
-                            break;
-                        },
-                        Err(SendError::Closed(v)) => {
-                            // The channel assigned to this wait group has been closed, so we will not
-                            // add it back to the list of wait groups, and we will try to send this
-                            // across another channel.
-                            df = v.0
-                        },
-                        Err(SendError::Full(_)) => unreachable!(),
+                for df in split_to_morsels(&df, ideal_morsel_size) {
+                    if raw_morsel_sender.send((df, morsel_seq)).await.is_err() {
+                        return Ok(());
                     }
-
-                    let Some(v) = wait_groups.next().await else {
-                        // All channels have closed
-                        break 'main;
-                    };
-
-                    indexed_wait_group = v;
+                    morsel_seq = morsel_seq.successor();
                 }
             }
-
-            // Join on the producer handle to catch errors/panics.
-            drop(df_stream);
-            metadata_task_handle.await.unwrap()
+            PolarsResult::Ok(())
         });
 
-        let raw_morsel_distributor_task_handle =
-            task_handles_ext::AbortOnDropHandle(raw_morsel_distributor_task_handle);
+        let join_task = io_runtime.spawn(async move {
+            metadata_task.await.unwrap()?;
+            prefetch_task.await.unwrap()?;
+            decode_task.await.unwrap()?;
+            distribute_task.await.unwrap()?;
+            Ok(())
+        });
 
-        (raw_morsel_receivers, raw_morsel_distributor_task_handle)
+        (raw_morsel_receivers, AbortOnDropHandle(join_task))
```

Comment:
Wrapping this `join_task` `AbortOnDropHandle` here doesn't cancel all of the previous tasks - the handles for those tasks should be individually wrapped in `AbortOnDropHandle` as soon as they are created.